src/KokkosPluginLoader.cpp

```diff
@@ -0,0 +1,145 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/util/KokkosPluginLoader.hpp"
+
+#ifndef _WIN32
+#include <dlfcn.h>
+#include <dirent.h>
+#endif
+
+const uint64_t kokkos_interface_version = 20171029;
+
+RAJA_INLINE
+bool
+isSharedObject(const std::string& filename)
+{
+  return (filename.size() > 3 && !filename.compare(filename.size() - 3, 3, ".so"));
+}
+
+template<typename function>
+RAJA_INLINE
+void
+getFunction(void* plugin, std::vector<function>& functions, const char* fname)
+{
+  #ifndef _WIN32
+  function func = (function) dlsym(plugin, fname);
+  if (func)
+    functions.push_back(func);
+  else
+    printf("[KokkosPluginLoader]: dlsym failed: %s\n", dlerror());
+  #else
+  RAJA_UNUSED_ARG(plugin);
+  RAJA_UNUSED_ARG(functions);
+  RAJA_UNUSED_ARG(fname);
+  #endif
+}
+
+namespace RAJA {
+namespace util {
+
+KokkosPluginLoader::KokkosPluginLoader()
+{
+  char *env = getenv("KOKKOS_PLUGINS");
+  if (env == nullptr)
+  {
+    return;
+  }
+  initDirectory(std::string(env));
+
+  for (auto &func : init_functions)
+  {
+    func(0, kokkos_interface_version, 0, nullptr);
+  }
+}
+
+void KokkosPluginLoader::preLaunch(RAJA::util::PluginContext& p)
+{
+  for (auto &func : pre_functions)
+  {
+    func("", 0, &(p.kID));
+  }
+}
+
+void KokkosPluginLoader::postLaunch(RAJA::util::PluginContext& p)
+{
+  for (auto &func : post_functions)
+  {
+    func(p.kID);
+  }
+}
+
+void KokkosPluginLoader::finalize()
+{
+  for (auto &func : finalize_functions)
+  {
+    func();
+  }
+}
+
+// Initialize plugin from a shared object file specified by 'path'.
+void KokkosPluginLoader::initPlugin(const std::string &path)
+{
+  #ifndef _WIN32
+  void *plugin = dlopen(path.c_str(), RTLD_NOW | RTLD_GLOBAL);
```

Comment:
Does this get dlclose()ed somehwere?

Reply:
Whoopsie, don't believe so actually. Adding it in should still I think, since the functions are already grabbed and stored in the vectors.

Reply:
I think it should only be called once we no longer need the library.